python/cugraph/cugraph/gnn/data_loading/dist_sampler.py

```diff
@@ -0,0 +1,349 @@
+            rank = minibatch_dict["rank"] if "rank" in minibatch_dict else 0
+
+            full_output_path = os.path.join(
+                self.__directory,
+                f"batch={rank:05d}.{start_batch_id:08d}-"
+                f"{rank:05d}.{end_batch_id:08d}.parquet",
+            )
+
+            results_dataframe_p.to_parquet(
+                full_output_path,
+                compression=None,
+                index=False,
+                force_nullable_schema=True,
+            )
+
+    def __write_minibatches_csr(self, minibatch_dict):
+        raise NotImplementedError(
+            "CSR format currently not supported for distributed sampling"
+        )
+
+    def write_minibatches(self, minibatch_dict):
+        if (minibatch_dict["majors"] is not None) and (
+            minibatch_dict["minors"] is not None
+        ):
+            self.__write_minibatches_coo(minibatch_dict)
+        elif (minibatch_dict["major_offsets"] is not None) and (
+            minibatch_dict["minors"] is not None
+        ):
+            self.__write_minibatches_csr(minibatch_dict)
+        else:
+            raise ValueError("invalid columns")
+
+
+class DistSampler:
+    def __init__(
+        self,
+        graph: Union[pylibcugraph.SGGraph, pylibcugraph.MGGraph],
+        writer: DistSampleWriter,
+        local_seeds_per_call: int = 32768,
+    ):
+        self.__graph = graph
+        self.__writer = writer
+        self.__local_seeds_per_call = local_seeds_per_call
+
+    def sample_batches(
+        self, seeds: TensorType, batch_ids: TensorType, random_state: int = 0
+    ):
+        raise NotImplementedError("Must be implemented by subclass")
+
+    def sample_from_nodes(
+        self, nodes: TensorType, *, batch_size: int = 16, random_state: int = 62
+    ):
+        batches_per_call = self._local_seeds_per_call // batch_size
+        actual_seeds_per_call = batches_per_call * batch_size
+
+        num_seeds = len(nodes)
+        nodes = torch.split(
+            torch.as_tensor(nodes, device="cuda"), actual_seeds_per_call
+        )
+
+        if self.is_multi_gpu:
+            rank = torch.distributed.get_rank()
+            world_size = torch.distributed.get_world_size()
+
+            t = torch.empty((world_size,), dtype=torch.int64, device="cuda")
+            local_size = torch.tensor(
+                [int(ceil(num_seeds / batch_size))], dtype=torch.int64, device="cuda"
+            )
+
+            torch.distributed.all_gather_into_tensor(t, local_size)
+            if (t != local_size).any() and rank == 0:
+                warnings.warn(
+                    "Not all ranks received the same number of batches. "
+                    "This might cause your training loop to hang due to uneven inputs."
+                )
+
+            batch_id_start = t.cumsum(dim=0)[rank] - t[0]
+        else:
+            batch_id_start = 0
+
+        for i, current_seeds in enumerate(nodes):
```

Comment:
Wont this be slow and  is this a singular node ? 
```suggestion
        for i, current_seed in enumerate(nodes):
```

Reply:
`current_seeds` is a view of all the seeds.  It contains the seeds for the current sampling call.  This is not slow.